src/TimeoutExtender.js

```diff
@@ -0,0 +1,163 @@
+    const now = Date.now()
+    this._addNode({
+      message,
+      receivedOn: now,
+      timerOn: now + this._visTimeout - API_CALL_LEAD_MS
+    })
+  }
+
+  /**
+   * Deletes a message from the tracker, if the message is currently being tracked.
+   * @param {Message} message A Squiss Message object
+   */
+  deleteMessage(message) {
+    const node = this._index[message.raw.MessageId]
+    if (node) this._deleteNode(node)
+  }
+
+  /**
+   * Adds a message wrapper node to the linked list and hash map index.
+   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
+   * object to be added
+   * @private
+   */
+  _addNode(node) {
+    this._index[node.message.raw.MessageId] = node
+    if (!this._head) {
+      this._head = node
+      this._tail = node
+      this._headChanged()
+    } else {
+      this._tail.next = node
+      node.prev = this._tail
+      this._tail = node
+    }
+  }
+
+  /**
+   * Deletes a message wrapper node from the linked list and hash map index.
+   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
+   * object to be removed
+   * @private
+   */
+  _deleteNode(node) {
+    const msgId = node.message.raw.MessageId
+    delete this._index[msgId]
+    if (this._head === node) {
+      this._head = node.next
+      if (this._head) delete this._head.prev
+      this._headChanged()
+    } else if (this._tail === node) {
+      this._tail = this._tail.prev
+      delete this._tail.next
+    } else {
+      node.prev.next = node.next
+      node.next.prev = node.prev
+    }
+  }
+
+  /**
+   * Called internally when the head of the linked list has changed in any way. This
+   * function is responsible for mantaining the timer that determines the tracker's next
+   * action.
+   * @returns {boolean} true if a timer was set in response to the changed head; false
+   * otherwise.
+   * @private
+   */
+  _headChanged() {
+    if (this._timer) clearTimeout(this._timer)
+    if (!this._head) return false
+    const node = this._head
+    this._timer = setTimeout(() => {
+      if (this._stopAfter) {
+        const age = Date.now() - node.receivedOn + API_CALL_LEAD_MS
+        if (age >= this._stopAfter) return this._deleteNode(node)
+      }
+      return this._renewNode(node)
+    }, node.timerOn - Date.now())
+    return true
+  }
+
+  /**
+   * Extends the VisbilityTimeout of the message contained in the provided wrapper node,
+   * and moves the node to the tail of the linked list.
+   * @param {{message: Message, receivedOn: number, timerOn: number}} node The node
+   * object to be renewed
+   * @private
+   */
+  _renewNode(node) {
+    this._squiss.changeMessageVisibility(node.message, this._visTimeout + API_CALL_LEAD_MS)
+      .then(() => this._squiss.emit('timeoutExtended', node.message))
+      .catch(err => {
+        this._squiss.emit('error', err)
+      })
+    this._deleteNode(node)
+    node.timerOn = Date.now() + this._visTimeout
```

Comment:
This is _already_ running 5 seconds earlier than the expiration due to subtracting `API_CALL_LEAD_MS` the first time. So by adding only `this._visTimeout`, I'm telling it to trigger the timer for this message 5 seconds before it expires the next time around.